File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, val=0, next=None):
#     self.val = val
#     self.next = next
# Definition for a binary tree node.
# class TreeNode:
#   def __init__(self, val=0, left=None, right=None):
#     self.val = val
#     self.left = left
#     self.right = right
class Solution:
  def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
    lst, curr = [], head
    while curr:
      lst.append(curr)
      curr = curr.next
    
    def helper(lb, ub):
      if lb == ub:
        return None
      mid = (lb+ub)//2
      return TreeNode(lst[mid].val, helper(lb, mid), helper(mid+1, ub))
    
    return helper(0, len(lst))
  

File: 0109-convert-sorted-list-to-binary-search-tree/0109-convert-sorted-list-to-binary-search-tree.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *   Val int
 *   Next *ListNode
 * }
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *   Val int
 *   Left *TreeNode
 *   Right *TreeNode
 * }
 */
func sortedListToBST(head *ListNode) *TreeNode {
  lst, curr := []*ListNode{}, head
  for curr != nil {
    lst = append(lst, curr)
    curr = (*curr).Next
  }
  return helper(0, len(lst), &lst)
}

func helper(lb int, ub int, lst *[]*ListNode) *TreeNode {
  if lb == ub {
    return nil
  }
  mid := (lb + ub) / 2
  return &TreeNode{
    Val: (*(*lst)[mid]).Val,
    Left: helper(lb, mid, lst),
    Right: helper(mid + 1, ub, lst),
  }
}